src/optimizer/statistics/stats_calculator.cpp

```diff
@@ -149,6 +149,58 @@ void StatsCalculator::Visit(const LogicalInnerJoin *op) {
   // TODO(boweic): calculate stats based on predicates other than join conditions
 }
 
+void StatsCalculator::Visit(const LogicalSemiJoin *op) {
+  // Check if there's join condition
+  NOISEPAGE_ASSERT(gexpr_->GetChildrenGroupsSize() == 2, "Join must have two children");
+  auto left_child_group = context_->GetMemo().GetGroupByID(gexpr_->GetChildGroupId(0));
+  auto right_child_group = context_->GetMemo().GetGroupByID(gexpr_->GetChildGroupId(1));
+  auto root_group = context_->GetMemo().GetGroupByID(gexpr_->GetGroupID());
+
+  // Calculate output num rows first
+  if (root_group->GetNumRows() == -1) {
+    size_t curr_rows = left_child_group->GetNumRows() * right_child_group->GetNumRows();
+    for (auto &annotated_expr : op->GetJoinPredicates()) {
+      // See if there are join conditions
+      if (annotated_expr.GetExpr()->GetExpressionType() == parser::ExpressionType::COMPARE_EQUAL &&
+          annotated_expr.GetExpr()->GetChild(0)->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE &&
+          annotated_expr.GetExpr()->GetChild(1)->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE) {
+        auto left_child = annotated_expr.GetExpr()->GetChild(0).CastManagedPointerTo<parser::ColumnValueExpression>();
+        auto right_child = annotated_expr.GetExpr()->GetChild(1).CastManagedPointerTo<parser::ColumnValueExpression>();
+        auto left_col = left_child->GetFullName();
+        auto right_col = right_child->GetFullName();
+        if ((left_child_group->HasColumnStats(left_col) && right_child_group->HasColumnStats(right_col)) ||
+            (left_child_group->HasColumnStats(right_col) && right_child_group->HasColumnStats(left_col))) {
+          curr_rows /= std::max(std::max(left_child_group->GetNumRows(), right_child_group->GetNumRows()), 1);
+        }
+      }
+    }
+    root_group->SetNumRows(static_cast<int>(curr_rows));
+  }
+
+  // TODO(elenac): For now we comment this because required_cols_ is always empty in the system
```

Comment:
Why is it always empty? When will it not always be empty? What's the timeline on all of this not being dead code? Next PR? Please elaborate this comment for people other than the original author.

Reply:
This is not always empty. I wrongly ignored the following case
https://github.com/cmu-db/noisepage/blob/f0a6e9f2cbc5ad3c9fe79846d9193806460b5403/src/optimizer/optimizer_task.cpp#L228
Uncommented this block to handle such cases. 

These test cases in nested-query.test covers it:
`SELECT pno FROM part WHERE pno in (SELECT pno FROM part WHERE pno in (SELECT shipment.pno FROM shipment where sno >= 1));
`
`
SELECT * FROM shipment WHERE pno <= (SELECT MAX(pno) FROM part) order by pno;
`